### PR TITLE
RFC: General 1-norm estimation for some functions of sparse matrices 

### PR DESCRIPTION
**Pull request by marcusps**
@mfasi @tkelman 

Here is what I had in mind for a more general `normest` routine.  At the moment there is something wrong with `Ac_ldiv_B!` that I don't quite understand (something about the types keeps this from compiling).

This attempt of a general `normest` is also not type stable (the `pre` routine in particular mucks that up) but that is probably fixable by replacing it with an object instead of a function.

I also put `normest` in a separate file just to make it easier to test it while I play with it.  This pull request is really just meant to bring attention to the idea and discuss the proper way to do it.

EDIT:

This is motivated by the usefulness of such a generic function for the fast and accurate computation of `f(A)*v` without explicitly computing `f(A)`, which is especially attractive when `A` is sparse but `f(A)` may not be. The one interesting case where this can be applied is `expm`, but potentially others as well, such as `logm`, `sqrtm` (see below).

_Computing the Action of the Matrix Exponential, with an Application to Exponential Integrators_, Awad H. Al-Mohy and Nicholas J. Higham, SIAM Journal on Scientific Computing 2011 33:2, 488-511. ([preprint](http://eprints.ma.man.ac.uk/1426/))

_Computing f(A)b for Matrix Functions f_, Philip I. Davies and Nicholas J. Higham. Lecture Notes in Computational Science and Engineering Volume 47, 2005, pp 15-24 ([preprint](http://eprints.ma.man.ac.uk/96/))

_Estimating the condition number of f(A)b_, Edvin Deadman. Numerical Algorithms December 2014 ([article](http://rd.springer.com/article/10.1007%2Fs11075-014-9947-4))

_An Improved Schur--Padé Algorithm for Fractional Powers of a Matrix and their Fréchet Derivatives_, Nicholas J Higham, Lijing Lin, SIAM. J. Matrix Anal. & Appl. 32 1056 2013. ([preprint](http://eprints.ma.man.ac.uk/2021/01/hili13.pdf))

### CODE REVIEW

**Comment by marcusps** [comment]
I have changed my approach slightly, and the generalization seems to work (this branch includes some minor changes to @mfasi's implementation of Higham and Tisseur's algorithm for norm estimation, but this first commit with the minor changes has been submited as a separate pull request).

I had to change how `cond()` and `normestinv()`  are tested, because the algorithm implemented is probabilistic and only provides upper and lower bounds to these two quantities.  Other than that, it should be easy to do 1 norm estimation for matrix powers (needed in ExpmV.jl) as well as other functions of sparse matrices.

File: test/sparsedir/sparse.jl [hunk 1]
```diff
@@ -1065,17 +1065,57 @@ Ai = ceil(Int,Ar*100)
...
+# test sparse matrix normestinv
+function rel_lb_err(a,b)
+    (b-a)/b
+end
+
+function rel_ub_err(a,b)
+    (a-b)/b
+end
+
+test_normestinv_Ac = Float64[begin
```
Comment (tkelman) [inline test/sparsedir/sparse.jl]: trailing whitespace on these - make sure they pass `make check-whitespace` locally

**Comment by marcusps** [comment]
@mfasi I have a fully working version now. I am curious to hear what you think.  There maybe a slight performance hit with `normestinv()` written this way, but I think it should be minor (although I have not measured).

**Comment by KristofferC** [comment]
Before you complicate the code with `@eval` I would suggest benchmarking to see if a significant overhead is induced. Maybe for smallish matrices it is relevant.

**Comment by tkelman** [comment]
Any updates here? Did other later PR's cover this functionality?

**Comment by marcusps** [comment]
No, not yet. I haven't had time to do careful benchmarking of this code vs what @mfasi suggested.

**Comment by marcusps** [comment]
I have some code elsewhere that does, but I have not delved into detailed
benchmarking yet — I just haven't had time.

**Comment by GiggleLiu** [comment]
why this pr is not merged yet?

**Comment by ViralBShah** [comment]
Would be nice to revive this, but it is also really old and may just need a fresh start.

**Comment by marcusps** [comment]
I'd be happy to see it revived — I don't think I have the bandwidth to get
back into it.

Indeed my motivation was to use this in sparse expmv, so I'm sure there
will be many users for it.

**Comment by ViralBShah** [comment]
Enough has changed since this was created that we'll need to substantially redo this if we want it.

**Pull request closed without merging**